### DISCUSSION TITLE
Unidirectional OneToOne relationship is not working

### DISCUSSION THREAD

**Discussion by aelfricraven** [jhipster/generator-jhipster#25901] [Q&A] [answered]
Hi,
I'm new to JHipster, but have seen a few examples already and it always seemd pretty straightforward.
Now that I'm giving it a try, I am already stumbling early on, with the JDL.
The documentation (https://www.jhipster.tech/jdl/relationships#examples) says I can create an unidirectional relationship by using this form:
`relationship OneToOne {`
`    Company{location} to Location`
`    Branch{location} to Location`
`}`
The generated code for location contains references to both Company and Branch, however. Am I missing something? My alternative woold be making the Location embeddable, but this is not possible, yet, if I understand it right.
Do you have any suggestions?
Thank!
    ciao Thomas

**Comment by Tcharl** [comment]
Hi,

The thing that you put into bracket is the name of the relationship. I guess that the framework miss a fatal exception here.
if you want to make it work:
```
relationship OneToOne {
    Company{location} to Location{company}
    Branch{location} to Location{branch}
}
```
Plus posting a bug on this issues tracker please ;-)

**Reply by aelfricraven** [reply]
Hi,
Hi,
just to make sure there's no misunderstanding: I don't want to have the Location class to have a reference to either the Company or the Branch. All I want is a unidirectional relationship. To achieve this (according to the documentation) you should leave the {} away in the Location end.
Your suggesstion would create a bidirectional relationship and ends up with both an branch and a company attribute inside of the location, which is not needed in my case (and clutters the code).
Thanks!
    ciao Thomas

**Comment by mshima** [comment, marked as answer]
Use ManyToOne instead of OneToOne.

OneToOne is bidirectional in JHipster because once the generator finds the OneToOne type on the left side, it creates the back reference. 
Constraints will be created in the back reference to make it a OneToOne relationship, otherwise it’s effectively a ManyToOne relationship.

**Reply by aelfricraven** [reply]
Hi,
thanks. Your suggestion generates the code I had in mind. I guess it's one of the quirks you must know in order to progress :)
   ciao Thomas

**Reply by mshima** [reply]
I recently had this problem working on https://github.com/jhipster/generator-jhipster/issues/22748.
It should be better documented.

**Reply by OmarHawk** [reply]
According to https://github.com/jhipster/generator-jhipster/blob/49658c0c929dd3bf1589961ed6f6cef2ed17c139/generators/server/support/relationship.ts#L33 it is required for filtering capabilities.

At the moment, I'm also trying to achieve a unidirectional OneToOne because somehow, hibernate is not actually LazyLoading that relationship in the one side where I need to get rid of EAGER loading, even though the annotation says "LAZY".  
A bit background here: 
I tried to introduce the OneToOne relationship due to performance reasons as it involves a large TextBlob field which makes my application quickly run out of memory in list views, so I do require a working lazy loading or alternatively a unidirectional relationship...

At the moment, I have just commented out the back reference manually, but I guess, I'm also looking into gettting rid of this somehow programmatically.

**Reply by mshima** [reply]
@OmarHawk supporting unidirectional OneToOne in JHipster can be tricky due to UI.
An alternative that may be easier to implement is:
```
relationship ManyToOne {
  @LiquibaseUnique Parent{child} to Child
}
```

**Reply by OmarHawk** [reply]
Ok, does not work for me, as I had previously something like this:

```
relationship OneToOne {
  Parent{child} to @Id @OnDelete("CASCADE") @OnUpdate("CASCADE") Child
}
```

which I converted to

```
relationship ManyToOne {
  @LiquibaseUnique Parent{child} to @Id @OnDelete("CASCADE") @OnUpdate("CASCADE") Child
}
```
as I wanted to avoid that extra column in the Child entity.

And that one does not work properly - the generated schema is dysfunctional as the generated liquibase relationship expects an additional column in the Child - which it does not have as it has the `@Id` annotation.

**Reply by OmarHawk** [reply]
What worked for me to make the specific relationship kind of unidirectional without producing inconsistencies was to just filter out the generated relationship through a local side-by-side blueprint like this:

```
get [BaseApplicationGenerator.POST_PREPARING_EACH_ENTITY]() {
    return this.asPostPreparingEachEntityTaskGroup({
      async postPreparingEachEntityTemplateTask({entity}) {
          entity.relationships = entity.relationships.filter(r => r.relationshipType !== 'one-to-one' || r.relationshipSide !== 'right');
      },
    });
  }
```

This way, the templates won't pick up this side of the relationship, making it effectively unidirectional in the way, I need it.
Don't know, if that's actually a good idea, but at least it works.

**Reply by mshima** [reply]
@OmarHawk I think we can remove the condition if it working and add backward compatibility.
Can you provide a PR?

**Reply by OmarHawk** [reply]
> @OmarHawk I think we can remove the condition if it working and add backward compatibility. Can you provide a PR?

Sure about the PR. Maybe, it would be good to have this behaviour configurable through some sort of annotation to not break / change existing generated applications, implicitly having backwards compatibility.

Something like 

```
relationship OneToOne {
  @Unidirectional Parent{child} to @Id @OnDelete("CASCADE") @OnUpdate("CASCADE") Child
}
```

**Reply by DanielFran** [reply]
@mshima @OmarHawk For simplicity, it would be better to add backward compatibility, so a OneToOne relationship is also a one to one in database side without the need of annotation.

**Reply by Hiisham89** [reply]
@OmarHawk I create a blueprint using below command and want to add this to my blueprint could you please tell me where i put this?
jhipster generate-blueprint --local-blueprint --sub-generators app --all-priorities

get [BaseApplicationGenerator.POST_PREPARING_EACH_ENTITY]() {
    return this.asPostPreparingEachEntityTaskGroup({
      async postPreparingEachEntityTemplateTask({entity}) {
          entity.relationships = entity.relationships.filter(r => r.relationshipType !== 'one-to-one' || r.relationshipSide !== 'right');
      },
    });
  }

**Reply by OmarHawk** [reply]
You should have a local side-by-side blueprint. 
In my case, my blueprint is extending the liquibase subgenerator of jhipster (as I also have a lot other stuff meanwhile, extending the functionality database-wise).

Inside that Blueprint should already be a _generator.mjs_ with an empty `postPreparingEachEntityTemplateTask`  method generated by your generate-blueprint command. This method you can just replace with the above mentioned code. Please be aware again that this would make _every_ one-to-one relationship unidirectional.
